### PR TITLE
Some API endpoints for articles

### PR DESCRIPTION
**Pull request by BaptisteGelez**
- Define the API with canapi
- Implement `canapi::Provider` for `Post` (only `get` and `list` actually)
- Use these functions to serve the API (`/api/v1/posts/<id>` and `/api/v1/posts`)
- Document the API with Swagger (couldn't test it, I hope it will be served correctly :confused:)

It is a first step for #5

### CODE REVIEW

File: plume-models/src/posts.rs [hunk 1]
```diff
@@ -55,6 +57,54 @@ pub struct NewPost {
...
+        let mut query = posts::table.into_boxed();
+        if let Some(title) = filter.title {
+            query = query.filter(posts::title.eq(title));
+        }
+        if let Some(subtitle) = filter.subtitle {
+            query = query.filter(posts::subtitle.eq(subtitle));
+        }
+        if let Some(content) = filter.content {
+            query = query.filter(posts::content.eq(content));
+        }
```
Comment (igalic) [inline plume-models/src/posts.rs]: what's the point of these queries and the override of the query variable?

Comment (igalic) [reply plume-models/src/posts.rs]: aahh, you're fetching different parts of the same article, and stuffing them into different boxes of the query variable!

File: plume-api/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+    ($url:expr => $ep:ty) => {
+        impl Endpoint for $ep {
+            type Id = i32;
+
+            fn endpoint() -> &'static str {
+                $url
+            }
+        }
+    };
+}
```
Comment (igalic) [inline plume-api/src/lib.rs]: sometimes i dream of lisp indentation style in Javascript. soon, I'll start dreaming of lisp indentation style in rust

File: po/gl.po [hunk 3]
```diff
@@ -325,7 +325,8 @@ msgstr[1] "{{ count }} autoras en este blog: "
 
 msgid "Login or use your Fediverse account to interact with this article"
 msgstr ""
-"Conéctese ou utilice a súa conta no fediverso para interactuar con este artigo"
+"Conéctese ou utilice a súa conta no fediverso para interactuar con este "
+"artigo"
```
Comment (igalic) [inline po/gl.po]: why does this file keep changing with every new build?

File: src/api/posts.rs [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+use plume_models::db_conn::DbConn;
+use plume_models::posts::Post;
+
+#[get("/posts/<id>")]
+fn get(id: i32, conn: DbConn) -> Json<serde_json::Value> {
+    let post = <Post as Provider<PgConnection>>::get(&*conn, id).ok();
+    Json(json!(post))
+}
+
+// TODO: handle query params
```
Comment (igalic) [inline src/api/posts.rs]: such as?

File: docs/api.yaml [hunk 5]
```diff
@@ -0,0 +1,45 @@
+openapi: "3.0"
```
[the remark this answers is not in the document]
Comment (BaptisteGelez) [reply docs/api.yaml]: not yet, but it may be possible thanks to canapi…

File: po/gl.po (hunk 3 above)
Comment (BaptisteGelez) [reply po/gl.po]: it is updated every time you run Plume. However, extracting i18n related task to an external tool may be better… :thinking:

File: src/api/posts.rs (hunk 4 above)
Comment (BaptisteGelez) [reply src/api/posts.rs]: For instance, if you `GET /api/v1/posts?title=Hello%20world`, only posts with the `Hello world` title should be shown (`title` is not the best example, but `author_id` could be interesting for instance)

File: plume-models/src/posts.rs (hunk 1 above)
Comment (BaptisteGelez) [reply plume-models/src/posts.rs]: actually, the `PostEndpoint` type is here to represent both the client's query and the result sent by the API. `filter` is what is requested by the client, so if any filter have been specified I update the SQL `query` to only return the requested posts.

Comment (BaptisteGelez) [reply plume-models/src/posts.rs]: not sure if I'm clear, tell me if not :grin:

**Comment by igalic** [comment]
i just realized how easy CLI mappings would be to implement, once the API is in place.

File: src/api/posts.rs (hunk 4 above)
Comment (fdb-hiroshima) [reply src/api/posts.rs]: So is this (at least partially) blocked by #149, or not related?

Comment (BaptisteGelez) [reply src/api/posts.rs]: No, actually we just need to get the raw query params, and deserialize them to `PostEndpoint` with something like `serde_qs`.

**Pull request merged**